### ISSUE TITLE
Implementing batching

### ISSUE THREAD

**Issue by ihavnoid** [leela-zero/leela-zero#410]
As an exercise to get some OpenCL and related programming experience, I looked at the code to see what we might need to do some batching (If I understood correctly, the 'batching' on the TODO list refers to 'grouping multiple NN evaluation to a single pipeline so that the GPU can do it more efficiently', right?).

As a first step, I implemented a simple pipeline with 4 batches by replacing the data to be float4 instead of float, plus other fixes here and there.  Got the same output, netbench showed 1/3 performance, but since we are doing 4 batches at once this means we are doing 33% more evaluations.

The next step I would want to get some insight is what should be changed to exploit this new pipeline.  I was thinking of creating a job queue on the OpenCL object, create one or two worker threads, where the worker thread groups 4 forward() calls and shoves it into the pipe.  This obviously will need a lot of threads doing the MCTS but I do not have enough insight on rest of the code and I would like to seek some advice.

(Obviously it isn't always possible to group into 4 forward() calls so we need some other ideas there, but I think we can do it by giving up grouping after a certain period)

**Comment by gcp** [comment]
This risks making the engine massively weaker at a fixed playout count, unfortunately. It can be used for self-play but you need to play multiple games at once and batch over the games, not searches from the same position.

Look at the pull requests where someone implemented a TCP server using other frameworks for the NN eval, which works that way.

For gameplay it might work or it might make the engine weaker. It pretty much needs testing if the massive loss in parallel search efficiency by blocking threads on each other is worth it. 

The main use of batching is that it allows large speedup by merging the DCNN evaluation and using techniques like Winograd.

**Comment by gcp** [comment]
>It pretty much needs testing if the massive loss in parallel search efficiency by blocking threads on each other is worth it.

Given that you proposed limiting the blocking to a limited amount of time, you can trade off the blocking time vs the speedup vs the loss in parallel search efficiency.

It's going to be fairly tricky (and fragile) to get something clearly stronger unless you take the next step to take advantage of the blocking.

**Comment by gcp** [comment]
>It pretty much needs testing if the massive loss in parallel search efficiency by blocking threads on each other is worth it.

One trick to try is to notice when the nodes that are queuing a position are the same. This means that: a) you shouldn't run em both b) the search is contending heavily along a critical path and will prefer latency over throughput.

You can also tune VIRTUAL_LOSS_COUNT.

**Comment by killerducky** [comment]
@ihavnoid is your batching change in a branch on your github clone? I'd like to take a look at it.

**Comment by ihavnoid** [comment]
@killerducky it isnt in a branch now.  The only thing I have is the opencl code itself, and there has been zero effort on doing any batching at this moment.  I wanted to get some insight on what might had to be done at this moment.

**Comment by ihavnoid** [comment]
I implemented the batching here:

https://github.com/ihavnoid/leela-zero/tree/nn_batch

Added command line argument --batchsize  (1 or 4)
--batchsize 4 evaluates four network evaluations concurrently on a single OpenCL call.  To use the concurrency, we need many threads (in my system it seems to saturate around 8)

On my i7-8700 + GTX1080, eight threads + batches of four shows at least 30% more evaluations/sec compared to 8 thread + batches of 1.  I am not sure how much strength this represents though.

(Using https://sjeng.org/zero/best_v1.txt.zip, 1330 evals/sec s. 980 evals/sec)

Will be happy to get more feedback and comments and get it eventually merged if it seems interesting.

**Comment by ihavnoid** [comment]
I updated my local branch furthermore:
-  https://github.com/ihavnoid/leela-zero/tree/nn_batch

Now it tries running batch sizes of 1, 2, 4 and 8, see which ones fails to compile/run, and then auto-adjusts its batch size depending on how many NN evaluations are piling up on the queue.

I did not submit a pull request as this is going to be a quite intrusive change that may affect other commits, hence it is likely that this is going to cause a helluva lot of merge conflicts.  Wanted to make sure that this is worth being merged in, code looks acceptable, and then I think we can see how to merge it in without breaking other in-progress work.

@gcp  - can you elaborate why this might make fixed play counts massively weaker?  I was thinking that the MCTS algorithm used here is inherently parallelizable, and this is not going to run all 1600 plays in parallel anyway.  This is just trying to parallelize 1, 2, 4 or 8 NN evaluations whenever possible.

**Comment by Akababa** [comment]
Sorry for asking a basic question here, but if the program doesn't use batches how could it be efficient on GPU? And what would batch normalization even do?

**Comment by ihavnoid** [comment]
AFAIK the concept of batches (when doing batch normalization) is only necessary when doing the training.  When playing the trained network, we just use the trained parameters and hence don't really need the batching

**Comment by ihavnoid** [comment]
Fixed a bug which generated corrupted results under wrong conditions - a vector was freed before it is guaranteed that the OpenCL engine picks up the input vector.

**Comment by Ttl** [comment]
I implemented OpenCL batching support for the Winograd implementation. I didn't add any support for batching different positions yet and it currently just duplicates the current position to fill the batch. 

However I don't see any benefit for the 192x15 network using Intel integrated GPU. 64x5 network does give about 30% more nodes/s. I don't have access to any better GPUs at the moment, but maybe it would work better with a faster GPU where overheads are relatively bigger?

I put the code at https://github.com/Ttl/leela-zero/tree/batching_test if someone with proper GPU wants to try it.

**Comment by gcp** [comment]
Maybe you can open a new issue? Might get lost here. We'll need to try a few implementations in the tree search how to best make use of it, if it's a gain.

**Comment by Ttl** [comment]
I'm currently traveling until July and don't have access to any faster computer than my laptop. Batching doesn't give any gain in performance with the recent networks and the integrated GPU on the laptop is so slow that I can't really do any statistically significant tests either. It didn't seem like worth it to try to implement it further at the moment and I just dumped the partially finished code here in case anyone else is interested.

**Comment by sethtroisi** [comment]
In active PR with other discussion closing old issue.
